module/skill/QuinticWalk/src/QuinticWalk.cpp

```diff
@@ -0,0 +1,383 @@
+                log<NUClear::WARN>(fmt::format("Robot roll exceeds threshold - {} > {}",
+                                               std::abs(RPY.x()),
+                                               current_config.imu_roll_threshold));
+                walk_engine.request_pause();
+            }
+            else if (std::abs(RPY.y()) > current_config.imu_pitch_threshold) {
+                log<NUClear::WARN>(fmt::format("Robot pitch exceeds threshold - {} > {}",
+                                               std::abs(RPY.y()),
+                                               current_config.imu_pitch_threshold));
+                walk_engine.request_pause();
+            }
+        });
+
+        on<Configuration>("quinticwalk.yaml").then([this](const Configuration& cfg) {
+            log_level = cfg["log_level"].as<NUClear::LogLevel>();
+
+            load_quintic_walk(cfg, normal_config);
+
+            // Make sure the walk engine has the parameters at least once
+            if (first_config) {
+                // Send these parameters to the walk engine
+                walk_engine.set_parameters(current_config.params);
+
+                imu_reaction.enable(current_config.imu_active);
+
+                first_config = false;
+            }
+        });
+
+        on<Configuration>("goalie/quinticwalk.yaml").then([this](const Configuration& cfg) {
+            load_quintic_walk(cfg, goalie_config);
+        });
+
+        on<Startup, Trigger<KinematicsModel>>().then("Update Kinematics Model", [this](const KinematicsModel& model) {
+            kinematicsModel = model;
+            first_run       = true;
+            current_orders.setZero();
+            is_left_support  = true;
+            falling          = false;
+            last_update_time = NUClear::clock::now();
+            walk_engine.reset();
+        });
+
+        // Runs every time the Walk provider starts (wasn't running)
+        on<Start<Walk>, With<Behaviour::State>>().then([this](const Behaviour::State& behaviour) {
+            walk_engine.reset();
+            if (behaviour == Behaviour::State::GOALIE_WALK) {
+                current_config = goalie_config;
+            }
+            else {
+                current_config = normal_config;
+            }
+            // Send these parameters to the walk engine
+            walk_engine.set_parameters(current_config.params);
+
+            imu_reaction.enable(current_config.imu_active);
+        });
+
+        // Runs every time the Walk task is removed from the director tree
+        on<Stop<Walk>>().then([this] {
+            const float dt = get_time_delta();
+            current_orders.setZero();
+            walk_engine.update_state(dt, current_orders);
+        });
+
+        // MAIN LOOP
+        on<Provide<Walk>,
+           Needs<LeftLegIK>,
+           Needs<RightLegIK>,
+           Every<UPDATE_FREQUENCY, Per<std::chrono::seconds>>,
```

Comment:
Rather than doing it this way, I think that it might be better to do a Trigger<Sensors> here and just update with each sensors frame, then you can do the IMU stuff in here too

Reply:
Do not update with sensors, it is far too slow for the walk engine.